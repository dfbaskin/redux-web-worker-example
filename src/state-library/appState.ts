export interface ColumnDefinition {
  id: string;
  width: number;
}

export interface ApplicationState {
  columns: ColumnDefinition[];
  data: any[][];
  formulas: (string | null)[];
  width: number;
  height: number;
  scrollLeft: number;
  scrollTop: number;
  usedColumnCount: number;
}

export const initialState: ApplicationState = {
  columns: [],
  data: [],
  formulas: [],
  width: 100,
  height: 100,
  scrollLeft: 0,
  scrollTop: 0,
  usedColumnCount: 0
};
