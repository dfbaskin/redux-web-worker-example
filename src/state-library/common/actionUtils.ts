import { Action } from "redux";

export interface PayloadAction<T> extends Action {
  payload: T;
}

export interface PayloadOrErrorAction<T> extends Action {
  payload: T | Error;
  error?: boolean;
}

export function emptyActionCreator(type: string): () => Action {
  const fn = function(): Action {
    return {
      type
    };
  };
  fn.toString = () => type;
  return fn;
}

export function payloadActionCreator<T>(
  type: string
): (payload: T) => PayloadAction<T> {
  const fn = function(payload: T): PayloadAction<T> {
    return {
      type,
      payload
    };
  };
  fn.toString = () => type;
  return fn;
}

export function payloadOrErrorActionCreator<T>(
  type: string
): (payload: T | Error) => PayloadOrErrorAction<T> {
  const fn = function(payload: T | Error): PayloadOrErrorAction<T> {
    return payload instanceof Error
      ? {
          type,
          payload,
          error: true
        }
      : {
          type,
          payload
        };
  };
  fn.toString = () => type;
  return fn;
}
