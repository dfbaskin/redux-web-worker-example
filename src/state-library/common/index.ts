export { Action } from "redux";
export * from "./reducerUtils";
export * from "./actionUtils";
