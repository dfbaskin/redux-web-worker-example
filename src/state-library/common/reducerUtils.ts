import { Action } from "redux";
import { produce, Draft } from "immer";

interface Reducer<T> {
  (draft: Draft<T>, action: Action): void;
}

type Definition = [Function, Function];

export function createReducer<T>(reducers: Definition[], initialState: T) {
  const reducerMap = new Map<string, Reducer<T>>(
    reducers.map(([ac, rd]) => [ac.toString(), rd as Reducer<T>])
  );
  return (state: T | undefined = initialState, action: Action): T => {
    const reducer = reducerMap.get(action.type);
    return reducer
      ? produce(state, draft => {
          reducer(draft, action);
        })
      : state;
  };
}
