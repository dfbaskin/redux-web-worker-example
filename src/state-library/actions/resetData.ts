import { PayloadAction, payloadActionCreator } from "../common";
import { ApplicationState } from "../appState";

interface DefaultColumnBase {
  width?: number;
  value?: (row: number) => any;
  formula?: string;
}

interface DefaultColumn extends DefaultColumnBase {
  columnIndex: number;
}

const DEFAULT_WIDTH = 80;

function randomBetween(min: number, max: number) {
  return Math.floor(Math.random() * (max - min)) + min;
}

const baseColumns: DefaultColumnBase[] = [
  {
    width: 50,
    value: row => row + 1
  },
  {
    value: () => randomBetween(100, 900)
  },
  {
    value: () => randomBetween(0, 1000)
  },
  {
    width: 100,
    value: () => randomBetween(100000, 999999)
  },
  {
    width: 150,
    value: () => Math.random()
  },
  {
    width: 100,
    formula: "C1 + C2 + C3"
  },
  {
    width: 150,
    formula: "C3 mile to km"
  },
  {
    width: 80,
    formula: "mean(C2, C3, C4)"
  }
];

export const defaultColumns = baseColumns.map(
  (details, columnIndex): DefaultColumn => ({
    ...details,
    columnIndex
  })
);

interface Payload {
  rows: number;
  cols: number;
}

export const resetDataAction = payloadActionCreator<Payload>("RESET_DATA");

export function resetDataReducer(
  draft: ApplicationState,
  action: PayloadAction<Payload>
): void {
  let { rows, cols } = action.payload;

  draft.columns = [];
  draft.formulas = [];
  draft.usedColumnCount = defaultColumns.length;

  for (const column of defaultColumns) {
    const idx = draft.columns.length + 1;
    draft.columns.push({
      id: `C${idx}`,
      width: column.width || DEFAULT_WIDTH
    });
    draft.formulas.push(column.formula || null);
  }
  while (draft.columns.length < cols) {
    const idx = draft.columns.length + 1;
    draft.columns.push({
      id: `C${idx}`,
      width: DEFAULT_WIDTH
    });
    draft.formulas.push(null);
  }

  draft.data = [];
  while (draft.data.length < rows) {
    const row = draft.data.length;
    const dataRow: any[] = [];
    for (let idx = 0; idx < cols; idx++) {
      if (idx < defaultColumns.length) {
        const { value } = defaultColumns[idx];
        dataRow.push(value ? value(row) : null);
      } else {
        dataRow.push(null);
      }
    }
    draft.data.push(dataRow);
  }
}
