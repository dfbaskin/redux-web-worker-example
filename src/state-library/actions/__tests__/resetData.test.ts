import { resetDataAction, resetDataReducer } from "../resetData";
import { ApplicationState, initialState } from "../../appState";
import { createReducer } from "../../common";

const reducer = createReducer<ApplicationState>(
  [[resetDataAction, resetDataReducer]],
  initialState
);

describe("reset data action", () => {
  it("should create default data", () => {
    const action = resetDataAction({
      rows: 0,
      cols: 0
    });
    const oldState: ApplicationState = initialState;
    const newState = reducer(oldState, action);
    expect(newState).toMatchSnapshot();
  });

  it("should create random data", () => {
    const action = resetDataAction({
      rows: 10,
      cols: 20
    });
    const oldState: ApplicationState = initialState;
    const newState = reducer(oldState, action);
    expect(newState.columns.length).toBe(20);
    expect(newState.data.length).toBe(10);
    for (let col = 0; col < 20; col++) {
      expect(newState.columns[col].id).toEqual(`C${col + 1}`);
    }
    for (let row = 0; row < 10; row++) {
      const dataRow = newState.data[row];
      expect(dataRow[0]).toEqual(row + 1);
    }
  });
});
