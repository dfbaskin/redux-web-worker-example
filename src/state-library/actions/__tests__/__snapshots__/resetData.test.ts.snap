// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reset data action should create default data 1`] = `
Object {
  "columns": Array [
    Object {
      "id": "C1",
      "width": 50,
    },
    Object {
      "id": "C2",
      "width": 80,
    },
    Object {
      "id": "C3",
      "width": 80,
    },
    Object {
      "id": "C4",
      "width": 100,
    },
    Object {
      "id": "C5",
      "width": 150,
    },
    Object {
      "id": "C6",
      "width": 100,
    },
    Object {
      "id": "C7",
      "width": 150,
    },
    Object {
      "id": "C8",
      "width": 80,
    },
  ],
  "data": Array [],
  "formulas": Array [
    null,
    null,
    null,
    null,
    null,
    "C1 + C2 + C3",
    "C3 mile to km",
    "mean(C2, C3, C4)",
  ],
  "height": 100,
  "scrollLeft": 0,
  "scrollTop": 0,
  "selectedColumnIndex": 5,
  "usedColumnCount": 8,
  "width": 100,
}
`;
